cura/Settings/MachineManager.py

```diff
@@ -1496,3 +783,268 @@ def setSettingForAllExtruders(self, setting_name: str, property_name: str, prope
+
+    def _setVariantNode(self, position, container_node):
+        self._global_container_stack.extruders[position].variant = container_node.getContainer()
+        self.activeVariantChanged.emit()
+
+    def _setGlobalVariant(self, container_node):
+        self._global_container_stack.variant = container_node.getContainer()
+
+    def _setMaterial(self, position, container_node = None):
+        if container_node:
+            self._global_container_stack.extruders[position].material = container_node.getContainer()
+        else:
+            self._global_container_stack.extruders[position].material = self._empty_material_container
+        # The _current_root_material_id is used in the MaterialMenu to see which material is selected
+        root_material_id = container_node.metadata["base_file"]
+        root_material_name = container_node.getContainer().getName()
+        if root_material_id != self._current_root_material_id[position]:
+            self._current_root_material_id[position] = root_material_id
+            self._current_root_material_name[position] = root_material_name
+            self.rootMaterialChanged.emit()
+
+    def activeMaterialsCompatible(self):
+        # check material - variant compatibility
+        if Util.parseBool(self._global_container_stack.getMetaDataEntry("has_materials", False)):
+            for position, extruder in self._global_container_stack.extruders.items():
+                if not extruder.material.getMetaDataEntry("compatible"):
+                    return False
+        return True
+
+    ## Update current quality type and machine after setting material
+    def _updateQualityWithMaterial(self):
+        current_quality = None
+        if self._current_quality_group:
+            current_quality = self._current_quality_group.quality_type
+        quality_manager = Application.getInstance()._quality_manager
+        candidate_quality_groups = quality_manager.getQualityGroups(self._global_container_stack)
+        available_quality_types = {qt for qt, g in candidate_quality_groups.items() if g.is_available}
+
+        if not self.activeMaterialsCompatible():
+            self._setEmptyQuality()
+            return
+
+        if not available_quality_types:
+            self._setEmptyQuality()
+            return
+
+        if current_quality in available_quality_types:
+            self._setQualityGroup(candidate_quality_groups[current_quality], empty_quality_changes = False)
+            return
+
+        quality_type = sorted(list(available_quality_types))[0]
+        preferred_quality_type = self._global_container_stack.getMetaDataEntry("preferred_quality_type")
+        if preferred_quality_type in available_quality_types:
+            quality_type = preferred_quality_type
+
+        self._setQualityGroup(candidate_quality_groups[quality_type], empty_quality_changes = True)
+
+    def _updateMaterialWithVariant(self, position: Optional[str]):
```

Comment:
Something for the material manager?

Reply:
Similar to `updateQualityWithMaterial`, this is like the replacement for the old `setActiveMaterial()`